packages/auto-tls/src/auto-tls.ts

```diff
@@ -0,0 +1,200 @@
+    this.started = true
+  }
+
+  async stop (): Promise<void> {
+    this.events.removeEventListener('self:peer:update', this.fetchCertificates)
+    this.shutdownController?.abort()
+    clearTimeout(this.renewTimeout)
+    await stop(this.fetchCertificates)
+    this.started = false
+  }
+
+  private _fetchCertificates (): void {
+    if (this.fetching || this.certificate != null) {
+      this.log('already fetching or already have a certificate')
+      return
+    }
+
+    const addresses = this.addressManager
+      .getAddresses()
+      .filter(ma => !isPrivate(ma) && !isLoopback(ma) && (
+        TCP.exactMatch(ma) ||
+        WebSockets.exactMatch(ma) ||
+        WebSocketsSecure.exactMatch(ma) ||
+        QUICV1.exactMatch(ma) ||
+        WebTransport.exactMatch(ma)
+      ))
+
+    if (addresses.length === 0) {
+      this.log('not fetching certificate as we have no public addresses')
+      return
+    }
+
+    this.fetching = true
+
+    this.fetchCertificate(addresses, {
+      signal: AbortSignal.timeout(this.timeout)
+    })
+      .catch(err => {
+        this.log.error('error fetching certificates %e', err)
+      })
+      .finally(() => {
+        this.fetching = false
+      })
+  }
+
+  private async fetchCertificate (mulitaddrs: Multiaddr[], options?: AbortOptions): Promise<void> {
+    this.log('fetching certificate')
+
+    // TODO: handle rate limit errors like "too many new registrations (10) from this IP address in the last 3h0m0s, retry after 2024-11-01 09:22:38 UTC: see https://letsencrypt.org/docs/rate-limits/#new-registrations-per-ip-address"
+
+    const base36EncodedPeer = base36.encode(this.peerId.toCID().bytes)
+    const domain = `${base36EncodedPeer}.${this.forgeDomain}`
+
+    // Create CSR
+    const [certificatePrivateKey, csr] = await acmeClient.forge.createCsr({
+      commonName: domain,
+      altNames: []
+    })
+
+    const accountPrivateKey = await acmeClient.forge.createPrivateKey()
+
+    const client = new acmeClient.Client({
+      directoryUrl: this.acmeDirectory,
+      accountKey: accountPrivateKey
+    })
+    const certString = await client.auto({
+      csr,
+      email: `${base36EncodedPeer}@libp2p.direct`,
+      termsOfServiceAgreed: true,
+      challengeCreateFn: async (authz, challenge, keyAuthorization) => {
+        const addresses = mulitaddrs.map(ma => ma.toString())
+
+        this.log('asking https://%s/v1/_acme-challenge to respond to the acme DNS challenge on our behalf', this.forgeEndpoint)
+        this.log('dialback public addresses: %s', addresses.join(', '))
+        const response = await this.clientAuth.authenticatedFetch(`https://${this.forgeEndpoint}/v1/_acme-challenge`, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify({
+            value: keyAuthorization,
+            addresses
+          }),
+          ...options
+        })
+
+        if (!response.ok) {
```

Comment:
Do we want to allow all 2xx responses?

Reply:
Yes, I live in hope that maybe one day we'll return a 204 for requests with no content instead of always 200s.